From aeab6373346c760eb1f3682302084f1573c32a98 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Mon, 18 Sep 2023 10:54:49 +0200
Subject: [PATCH] dmaengine: stm32-mdma: correct desc prep when channel running

In case of the prep descriptor while the channel is already running
the CCR register value stored into the channel could already have
its EN bit set.  This would lead to a bad transfer since, at
start transfer time, enabling the channel while other registers
aren't yet properly set.
To avoid this, ensure to mask the CCR_EN bit when storing the
ccr value into the mdma channel structure.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Change-Id: I482fb5c591f0c489a32136d4a453dc7815646ad2
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/328082
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/dma/stm32-mdma.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: kernel-source/drivers/dma/stm32-mdma.c
===================================================================
--- kernel-source.orig/drivers/dma/stm32-mdma.c
+++ kernel-source/drivers/dma/stm32-mdma.c
@@ -490,7 +490,7 @@ static int stm32_mdma_set_xfer_param(str
 	src_maxburst = chan->dma_config.src_maxburst;
 	dst_maxburst = chan->dma_config.dst_maxburst;
 
-	ccr = stm32_mdma_read(dmadev, STM32_MDMA_CCR(chan->id));
+	ccr = stm32_mdma_read(dmadev, STM32_MDMA_CCR(chan->id)) & ~STM32_MDMA_CCR_EN;
 	ctcr = stm32_mdma_read(dmadev, STM32_MDMA_CTCR(chan->id));
 	ctbr = stm32_mdma_read(dmadev, STM32_MDMA_CTBR(chan->id));
 
@@ -966,7 +966,7 @@ stm32_mdma_prep_dma_memcpy(struct dma_ch
 	if (!desc)
 		return NULL;
 
-	ccr = stm32_mdma_read(dmadev, STM32_MDMA_CCR(chan->id));
+	ccr = stm32_mdma_read(dmadev, STM32_MDMA_CCR(chan->id)) & ~STM32_MDMA_CCR_EN;
 	ctcr = stm32_mdma_read(dmadev, STM32_MDMA_CTCR(chan->id));
 	ctbr = stm32_mdma_read(dmadev, STM32_MDMA_CTBR(chan->id));
 	cbndtr = stm32_mdma_read(dmadev, STM32_MDMA_CBNDTR(chan->id));
